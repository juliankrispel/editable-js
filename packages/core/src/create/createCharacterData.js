// @flow

import type { RawBlock, CharacterData } from '../types'

export default function createCharacterData(block: RawBlock): Array<CharacterData> {
  if (block.text != null) {
    return block.text.split('').map((_, index) => {
      const data = {
        marks: []
      }

      if (block.characterRanges != null) {
        block.characterRanges.filter(range => {
          return range.offset <= index &&
          range.offset + range.length > index
        }).forEach(range => {
          if (Array.isArray(range.marks)) {
            data.marks = data.marks.concat(range.marks)
          }

          if (range.entity != null) {
            data.entity = range.entity
          }
        })
      }

      return data
    })
  }
  return []
}
