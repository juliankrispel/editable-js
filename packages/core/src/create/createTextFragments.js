// @flow
import type { TextFragment, Block, EntityMap } from '../types'
import hasEqualCharacterData from '../hasEqualCharacterData'

export default function createTextFragments(block: Block, entityMap: EntityMap): Array<TextFragment> {
  return block.characterData.reduce(
    (acc, data, index) => {
      if (acc.length < 1) {
        return [{
          marks: data.marks,
          entity: data.entity,
          offset: index,
          text: block.text[index]
        }]
      } else {
        const lastFragment = acc[acc.length - 1]
        if (hasEqualCharacterData(lastFragment, data)) {
          return acc.slice(0, -1).concat([{
            ...lastFragment,
            text: lastFragment.text + block.text[index]
          }])
        } else {
          return [
            ...acc,
            {
              marks: data.marks,
              entity: data.entity,
              offset: index,
              text: block.text[index]
            }
          ]
        }
      }
    },
    []
  ).map(fragment => ({
    ...fragment,
    entity: fragment.entity != null ? entityMap[fragment.entity] : null
  }))
}
