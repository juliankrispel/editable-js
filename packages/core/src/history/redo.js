// @flow

import produce, { applyPatches } from 'immer'
import type { EditorState } from '../types'

const redo = (editorState: EditorState): EditorState => {
  const { changeIndex, changes } = editorState
  if (changes[changeIndex] != null) {
    let newState = applyPatches(editorState, changes[changeIndex].forward)
    return produce(newState, draft => {
      draft.lastCommitted = 'redo'
      draft.changeIndex++
    })
  } else {
    return editorState
  }
}

export default redo
