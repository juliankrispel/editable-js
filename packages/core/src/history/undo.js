// @flow

import produce, { applyPatches } from 'immer'
import type { EditorState } from '../types'

const undo = (editorState: EditorState, suppressCommit?: boolean): EditorState => {
  const { changeIndex, changes } = editorState

  if (changeIndex > 0) {
    let newState = applyPatches(editorState, changes[changeIndex - 1].reverse)

    return produce(newState, draft => {
      if (!suppressCommit) {
        draft.lastCommitted = 'undo'
      }
      draft.changeIndex--
    })
  } else {
    return editorState
  }
}

export default undo
