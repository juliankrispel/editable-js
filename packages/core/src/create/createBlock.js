// @flow

import type { RawBlock, Block } from '../types'
import createId from './createId'
import createCharacterData from './createCharacterData'

export default function createBlock (block: RawBlock): Block {
  const { children, text = '', key, characterRanges, ...rest } = block

  if (process.env.NODE_ENV !== 'production' && key != null && !isNaN(Number(key))) {
    throw new Error('block key can not be a number')
  }

  const newBlock = {
    ...rest,
    text,
    characterData: createCharacterData(block),
    children: [],
    key: key || createId()
  }

  if (Array.isArray(children)) {
    newBlock.children = children.map(createBlock)
  }

  return newBlock
}
