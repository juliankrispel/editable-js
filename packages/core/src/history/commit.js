// @flow

import produce, { applyPatches } from 'immer'

import undo from './undo'
import type { EditorState } from '../types'

const mergeLastCommit = (editorState: EditorState) => {
  const prevEditorState = undo(undo(editorState, true), true)
  let changes = { forward: [], reverse: [] }
  const lastTwoChanges = editorState.changes.slice(-2)

  const newEditorState = produce(
    prevEditorState,
    draft => {
      lastTwoChanges.forEach(change => {
        // console.log('change', JSON.stringify(change))
        applyPatches(draft, change.forward)
      })
    }, (forward, reverse) => {
      changes = {forward, reverse}
    }
  )

  return produce(
    newEditorState,
    draft => {
      draft.changes.pop()
      draft.changes[draft.changes.length - 1] = changes
      draft.changeIndex = draft.changes.length
    }, () => {}
  )
}

const commit = (
  editorState: EditorState,
  update: (EditorState, ...args:Array<any>) => void,
  ...rest: Array<any>
) => {
  let changes = { forward: [], reverse: [] }

  const { lastCommitted } = editorState

  let newEditorState = produce(
    editorState,
    draft => {
      update(draft, ...rest)
    },
    (forward, reverse) => {
      changes = { forward, reverse }
    }
  )

  newEditorState = produce(
    newEditorState,
    draft => {
      const { changeIndex } = draft

      draft.changeIndex++

      if (draft.changes.length > changeIndex) {
        draft.changes = draft.changes.slice(0, changeIndex - 1)
      }

      draft.changes.push(changes)

      draft.lastCommitted = update.name
    }
  )

  if (lastCommitted === update.name) {
    newEditorState = mergeLastCommit(newEditorState)
  }

  return newEditorState
}

export default commit
