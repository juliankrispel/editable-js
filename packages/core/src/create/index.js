import createBlock from './createBlock'
import createId from './createId'
import createCharacterData from './createCharacterData'
import createTextFragments from './createTextFragments'
import createEditorState from './createEditorState'

export {
  createBlock,
  createEditorState,
  createTextFragments,
  createCharacterData,
  createId
}
