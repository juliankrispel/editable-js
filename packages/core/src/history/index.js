import commit from './commit'
import redo from './redo'
import undo from './undo'

export {
  commit,
  redo,
  undo
}
