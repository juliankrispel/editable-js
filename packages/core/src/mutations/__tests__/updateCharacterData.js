// @flow

import { createEditorState } from '../../create'
import { commit } from '../../history'
import updateCharacterData from '../updateCharacterData'

describe('updateCharacterData', () => {
  test('updates character data at given range', () => {
    const initialState = createEditorState({
      blocks: [{
        key: '1',
        text: 'One',
        children: [{
          key: '2',
          text: 'Two',
          children: [{
            key: '3',
            text: 'Three'
          }]
        }]
      }]
    })

    const selection = {
      startKey: '1',
      endKey: '3',
      startOffset: 2,
      endOffset: 3
    }

    const result = commit(
      initialState,
      updateCharacterData,
      selection,
      (data) => {
        data.marks.push('bold')
      }
    )

    expect(result).toMatchSnapshot()
  })
})
