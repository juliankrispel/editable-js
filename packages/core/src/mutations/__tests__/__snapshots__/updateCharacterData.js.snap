// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateCharacterData updates character data at given range 1`] = `
Object {
  "changeIndex": 1,
  "changes": Array [
    Object {
      "forward": Array [
        Object {
          "op": "add",
          "path": Array [
            "content",
            0,
            "characterData",
            2,
            "marks",
            0,
          ],
          "value": "bold",
        },
        Object {
          "op": "add",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "characterData",
            0,
            "marks",
            0,
          ],
          "value": "bold",
        },
        Object {
          "op": "add",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "characterData",
            1,
            "marks",
            0,
          ],
          "value": "bold",
        },
        Object {
          "op": "add",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "characterData",
            2,
            "marks",
            0,
          ],
          "value": "bold",
        },
        Object {
          "op": "add",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "children",
            0,
            "characterData",
            0,
            "marks",
            0,
          ],
          "value": "bold",
        },
        Object {
          "op": "add",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "children",
            0,
            "characterData",
            1,
            "marks",
            0,
          ],
          "value": "bold",
        },
        Object {
          "op": "add",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "children",
            0,
            "characterData",
            2,
            "marks",
            0,
          ],
          "value": "bold",
        },
      ],
      "reverse": Array [
        Object {
          "op": "replace",
          "path": Array [
            "content",
            0,
            "characterData",
            2,
            "marks",
            "length",
          ],
          "value": 0,
        },
        Object {
          "op": "replace",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "characterData",
            0,
            "marks",
            "length",
          ],
          "value": 0,
        },
        Object {
          "op": "replace",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "characterData",
            1,
            "marks",
            "length",
          ],
          "value": 0,
        },
        Object {
          "op": "replace",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "characterData",
            2,
            "marks",
            "length",
          ],
          "value": 0,
        },
        Object {
          "op": "replace",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "children",
            0,
            "characterData",
            0,
            "marks",
            "length",
          ],
          "value": 0,
        },
        Object {
          "op": "replace",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "children",
            0,
            "characterData",
            1,
            "marks",
            "length",
          ],
          "value": 0,
        },
        Object {
          "op": "replace",
          "path": Array [
            "content",
            0,
            "children",
            0,
            "children",
            0,
            "characterData",
            2,
            "marks",
            "length",
          ],
          "value": 0,
        },
      ],
    },
  ],
  "content": Array [
    Object {
      "characterData": Array [
        Object {
          "marks": Array [],
        },
        Object {
          "marks": Array [],
        },
        Object {
          "marks": Array [
            "bold",
          ],
        },
      ],
      "children": Array [
        Object {
          "characterData": Array [
            Object {
              "marks": Array [
                "bold",
              ],
            },
            Object {
              "marks": Array [
                "bold",
              ],
            },
            Object {
              "marks": Array [
                "bold",
              ],
            },
          ],
          "children": Array [
            Object {
              "characterData": Array [
                Object {
                  "marks": Array [
                    "bold",
                  ],
                },
                Object {
                  "marks": Array [
                    "bold",
                  ],
                },
                Object {
                  "marks": Array [
                    "bold",
                  ],
                },
                Object {
                  "marks": Array [],
                },
                Object {
                  "marks": Array [],
                },
              ],
              "children": Array [],
              "key": "3",
              "text": "Three",
            },
          ],
          "key": "2",
          "text": "Two",
        },
      ],
      "key": "1",
      "text": "One",
    },
  ],
  "currentCharacterData": Object {
    "entity": null,
    "marks": Array [],
  },
  "entityMap": undefined,
  "lastCommitted": "updateCharacterData",
  "selection": Object {
    "endKey": "1",
    "endOffset": 0,
    "startKey": "1",
    "startOffset": 0,
  },
}
`;
